Add large diagonal flat to steep and dive loop track pieces

X123M3-256 opened the pull request:
This PR adds two new track pieces, a diagonal large steep to flat and a "dive loop", which is a diagonal steep to inversion piece that is primarily intended for building dive loops though there are several other useful elements that can be made with it.

![divelooppr](https://github.com/user-attachments/assets/1cab2791-ccf2-4517-b6c5-7c6df6bac8f1)

Track sprites are provided for the LSM launched coaster, single rail coaster, twister coaster, and corkscrew coaster (and of course other ride types that share those sprites). As I expect this PR to take a while to get looked at, I may add additional track sprites if I have them ready before this is merged, but I consider this ready for review in its current state.

Three new vehicle sprite groups were added for the dive loop track pieces. I will shortly submit PRs to the object repo to update the official train types with these sprites. It amounts to 128 new sprites (not including peep sprites) which is less than were added for the zero G rolls, so I think that's reasonable. The diagonal large flat to steep pieces do not require any new vehicle sprites and have been made available by default on all ride types for which track sprites are available, except those that get small flat to steep due to UI limitations.

Gymnasiast commented:
@SpartanFrederic104 @ZeeMaji could you two take these for a test?

ZeeMaji commented:
Sure after i test the holding brakes fix. Could we have some vehicles to test with?

SpartanFrederic104 commented:
![Fred's Workbench 2025-01-08 10-53-31](https://github.com/user-attachments/assets/327c51e4-9200-40c8-a790-c0c34dd7a50a)
![Fred's Workbench 2025-01-08 10-51-49](https://github.com/user-attachments/assets/a6b8bd7a-f017-4136-bc9e-f4d6ee4ed6fe)
Doing some testing and so far the new track sprites look great, but here are some issues I discovered:

![Fred's Workbench 2025-01-08 11-10-04](https://github.com/user-attachments/assets/5342b71c-3676-487a-9844-3375f5f4bc71)
![Fred's Workbench 2025-01-08 11-16-59](https://github.com/user-attachments/assets/b1fc796b-a868-4083-892d-edaeefb5b525)
1. Apparently you can make the large diagonal flat-to-steep pieces have a chain lift on the single rail, (hyper-)twister, and vertical drop coasters with the "Enable all drawable track pieces" cheat enabled (applies to both the flat-to-steep and steep-to-flat pieces). Of course they don't have sprites for chainlifts, yet they still function like a chain lift anyway. I believe the option to make the large diagonal flat-to-steep/steep-to-flat pieces have chainlifts should be disabled.

![Fred's Workbench 2025-01-08 11-09-12](https://github.com/user-attachments/assets/520c27e7-c3c6-4ee6-8aa8-101cda9d0cca)
2. The flying and lay-down coaster currently do not have access to large diagonal flat-to-steep pieces when the track is in the upright (e.g. laying-down) position.

X123M3-256 commented:
> Could we have some vehicles to test with from x7? The currently packed vehicles in the object release this uses won't have sprites for these i presume.

Sorry for the delay, I didn't think it would take me this long but I have submitted a PR to the object repo here https://github.com/OpenRCT2/objects/pull/372, which includes updated vehicles for the hybrid, single rail, and LSM launched coaster trains (no track sprites are available for the hybrid yet but I thought I might as well update the vehicle anyway).

I think you will need to replace your existing copy of those objects when testing with these - if it's easier to test with a .parkobj I can send you an updated hydraulic launch coaster train.

Gymnasiast commented:
> Apparently you can make the large diagonal flat-to-steep pieces have a chain lift on the single rail, (hyper-)twister, and vertical drop coasters with the "Enable all drawable track pieces" cheat enabled (applies to both the flat-to-steep and steep-to-flat pieces). Of course they don't have sprites for chainlifts, yet they still function like a chain lift anyway. I believe the option to make the large diagonal flat-to-steep/steep-to-flat pieces have chainlifts should be disabled.

I would rather just have sprites for large flat-to-steep diagonal pieces with chainlift, even if they’re hidden behind cheats normally.

SpartanFrederic104 commented:
I will actually agree with that. They should be available, but only behind cheats.

karst commented:
> I would rather just have sprites for large flat-to-steep diagonal pieces with chainlift, even if they’re hidden behind cheats normally.

I think that's a great idea, but also beyond the scope of this PR as these chain lifts are also not available on the regular version of it as of yet.

X123M3-256 commented:
> I would rather just have sprites for large flat-to-steep diagonal pieces with chainlift, even if they’re hidden behind cheats normally.

I can, if desired, provide lift sprites for the single rail and giga track but I'd really rather not have to make lift sprites for the other two (those are animated, they would have to be hand drawn). Also, this would not be consistent with the regular large steep to flat pieces which don't currently get lift sprites either - if I add lift sprites to these pieces, I should really add them to the existing large steep to flat and on the giga track verticals would probably be good as well. So I think if that needs to be done it should be a separate PR.

Also, even though this issue currently only shows with "enable all drawable track pieces" enabled, that is only because the ride types that get steep lifts have small steep to flat by default. If these pieces were added to the hybrid coaster, it would let you put lifts on these pieces without cheats, so this is still a bug that needs to be fixed even if those sprites are added, unless you want them available by default.

I must admit that I was a bit confused by the UI code so I'm not surprised I missed something.

Gymnasiast commented:
Very well, let’s leave that for another time.

Let me know if you need help with the UI bits.

X123M3-256 commented:
I have rebased the branch and fixed the issue with the lift hills. It was actually nothing to do with the UI code, I had accidentally included the ALLOW_CHAIN_LIFT flag on the large flat to steep pieces because I'd copied the flags from the small flat to steep. However, there is a secondary issue - even though the chain lift can no longer be selected, the chain lift button is still displayed. This also affects the regular large flat to steep pieces so I didn't try and fix that, but I can do that if needed.

I have also added the large diagonal flat to steep pieces to the lay down and flying coaster as requested. For some reason, the large flat to steep is available by default on the lay down but only with cheats for the flying coaster. I don't know why that is but I have kept the diagonal pieces consistent with the existing ones.

ZeeMaji commented:
I think the BM flyer additions may have been made before we made the large flat to steep pieces universal on all tracks that can draw them so that might be why it's in cheats. I would just move it out of cheats although that could be done separately from this PR after it's merged.

ZeeMaji commented:
After a discussion in the discord, i believe the dive loops should be made cheat-only on the corkscrew coaster as no real world arrow or old school vekoma coaster (that i'm aware of) has an element that resembles them. Tennessee Tornado's sidewinder *somewhat* resembles it but i feel like you can just recreate it better with the large corkscrew.

X123M3-256 commented:
> After a discussion in the discord, i believe the dive loops should be made cheat-only on the corkscrew coaster as no real world arrow or old school vekoma coaster (that i'm aware of) has an element that resembles them.

Done

Gymnasiast commented:
@ZeeMaji @SpartanFrederic104 Could you approve the PR if all your points have been addressed?

SpartanFrederic104 reviewed (approved):
New dive loop and diagonal large steep slope pieces look great, and old saves work fine. All my points have been addressed. So far so good. Don't forget to add a changelog entry by the way.

As for the flying coaster's large steep slope pieces (in the upright track position) being made available by default (i.e. without cheats), that can indeed be saved for another PR.

ZeeMaji reviewed (approved):
Everything seems to be about what i'd expect gameplay wise in regards to the pieces themselves and their availability, and i haven't noticed any issues with the sprites which all look lovely.
